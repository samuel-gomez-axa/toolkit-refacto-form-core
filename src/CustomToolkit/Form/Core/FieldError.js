import React from 'react';
import PropTypes from 'prop-types';
import { ClassManager } from '@axa-fr/react-toolkit-core';
import MessageTypes from './MessageTypes';

const propTypes = {
  message: PropTypes.string,
  messageType: PropTypes.oneOf([
    MessageTypes.error,
    MessageTypes.success,
    MessageTypes.warning,
  ]),
};

const defaultProps = {
  message: null,
  messageType: MessageTypes.error,
};

const defaultClassName = 'af-form__message';

const FieldError = ({ message, messageType }) => {
  const className = ClassManager.getComponentClassName(
    defaultClassName,
    messageType,
    defaultClassName
  );

  const classNameError = `af-form__${messageType}-text`;

  return message ? (
    <small className={className}>
      <span className="glyphicon glyphicon-exclamation-sign" />
      <span className={classNameError}>{message}</span>
    </small>
  ) : null;
};

FieldError.propTypes = propTypes;
FieldError.defaultProps = defaultProps;

FieldError.Clone = (data) => {
  const { props, message, child } = data;
  return React.cloneElement(child, {
    ...props,
    message,
  });
};

export default FieldError;
