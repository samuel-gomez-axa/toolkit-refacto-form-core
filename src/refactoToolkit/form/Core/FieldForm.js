import React, { useState, useRef, useEffect } from "react";
import PropTypes from "prop-types";
import { ClassManager } from "@axa-fr/react-toolkit-core";
import MessageTypes from "./MessageTypes";

const defaultClassName = "md-10";

const INITIAL_STATE = {
  hasLostFocusOnce: false,
  hasFocus: false,
  hasChange: false,
  memory: { message: "", messageType: "" },
};

export const FieldForm = (props) => {
  const {
    message,
    messageType,
    children,
    className,
    classModifier,
    forceDisplayMessage,
  } = props;
  const [state, setState] = useState(INITIAL_STATE);
  const { hasChange } = state;

  const previousForceDisplayMessage = useRef(forceDisplayMessage);

  useEffect(() => {
    if (previousForceDisplayMessage.current !== forceDisplayMessage) {
      setState((prevState) => ({
        ...prevState,
        memory: {
          message,
          messageType,
        },
      }));
    }
  }, [forceDisplayMessage, message, messageType]);

  const onChange = () => {
    // for particular case on particular browers which does not throw onFocus or onBlur event
    if (!hasChange) {
      setState((prevState) => ({ ...prevState, hasChange: true }));
    }
  };

  const onBlur = () =>
    setState((prevState) => ({
      ...prevState,
      hasLostFocusOnce: true,
      hasFocus: false,
    }));

  const onFocus = () =>
    setState((prevState) => ({
      ...prevState,
      hasFocus: true,
      memory: { message, messageType },
    }));

  const childrenCloned = renderedChildren({
    children,
    wrapper: { onFocus, onChange, onBlur },
    ...getInfo({ ...state, ...props }),
  });

  const subComponentClassName = ClassManager.getComponentClassName(
    className,
    classModifier,
    defaultClassName,
  );

  return <div className={subComponentClassName}>{childrenCloned}</div>;
};

FieldForm.propTypes = {
  children: PropTypes.node.isRequired,
  forceDisplayMessage: PropTypes.bool,
  message: PropTypes.string,
  messageType: PropTypes.string,
  className: PropTypes.string,
  classModifier: PropTypes.string,
};

FieldForm.defaultProps = {
  forceDisplayMessage: false,
  message: null,
  messageType: MessageTypes.error,
  className: defaultClassName,
  classModifier: null,
};

export const getInfo = ({
  forceDisplayMessage,
  hasLostFocusOnce,
  hasFocus,
  memory,
  hasChange,
  message,
  messageType,
}) => {
  const isDisplayMessage =
    hasLostFocusOnce || forceDisplayMessage || (hasChange && !hasFocus);

  if (!isDisplayMessage) {
    return {
      message: "",
      messageType: "",
    };
  }
  return hasFocus
    ? memory
    : {
        message,
        messageType,
      };
};

export const renderedChildren = ({ children, wrapper, message, messageType }) =>
  !children
    ? null
    : React.Children.map(children, (child) => {
        // create a copy that includes additional property values as needed

        if (!child || !child.props || !child.type) {
          return child;
        }

        const props = {
          ...child.props,
        };

        if (child.props.children) {
          const subChildren = renderedChildren({
            children: child.props.children,
            wrapper,
            message,
            messageType,
          });

          if (subChildren !== null) {
            props.children = subChildren;
          }
        }

        return child.type.Clone
          ? child.type.Clone({
              props,
              messageType,
              message,
              child,
              wrapper,
            })
          : React.cloneElement(child, props);
      });

export default FieldForm;
