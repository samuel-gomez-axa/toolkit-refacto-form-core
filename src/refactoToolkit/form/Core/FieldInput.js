import React from "react";
import PropTypes from "prop-types";
import { ClassManager } from "@axa-fr/react-toolkit-core";
import MessageTypes from "./MessageTypes";

const propTypes = {
  children: PropTypes.node.isRequired,
  childrenRight: PropTypes.node,
  helpMessage: PropTypes.string,
  message: PropTypes.string,
  messageType: PropTypes.oneOf([
    MessageTypes.error,
    MessageTypes.success,
    MessageTypes.warning,
  ]),
  isVisible: PropTypes.bool,
  className: PropTypes.string,
  classModifier: PropTypes.string,
};

const defaultClassName = "";
const defaultProps = {
  helpMessage: null,
  childrenRight: null,
  message: null,
  messageType: MessageTypes.error,
  isVisible: true,
  className: defaultClassName,
  classModifier: "",
};

const FieldInput = (props) => {
  const { children, className, classModifier, isVisible } = props;
  if (!isVisible) {
    return false;
  }
  const newClassName = ClassManager.getComponentClassName(
    className,
    classModifier,
    defaultClassName,
  );
  return <div className={newClassName}>{children}</div>;
};

FieldInput.propTypes = propTypes;
FieldInput.defaultProps = defaultProps;

export default FieldInput;
