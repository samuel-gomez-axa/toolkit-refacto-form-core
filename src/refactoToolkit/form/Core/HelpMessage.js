import React from "react";
import PropTypes from "prop-types";

const propTypes = {
  message: PropTypes.string,
  isVisible: PropTypes.bool,
};

const defaultProps = {
  message: null,
  isVisible: true,
};

const HelpMessage = ({ message, isVisible }) => {
  if (isVisible) {
    return <small className="af-form__help">{message}</small>;
  }
  return null;
};

HelpMessage.propTypes = propTypes;
HelpMessage.defaultProps = defaultProps;

export default HelpMessage;
