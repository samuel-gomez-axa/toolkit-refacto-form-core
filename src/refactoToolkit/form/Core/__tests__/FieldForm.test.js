import React, { Fragment } from "react";
import { render, within } from "@testing-library/react";
import FieldForm, { renderedChildren, getInfo } from "../FieldForm";
import FieldError from "../FieldError";
import FieldInput from "../FieldInput";
import HelpMessage from "../HelpMessage";

const CustomWithClone = ({ label, message, ...rest }) => (
  <p {...rest}>
    {label} {message}
  </p>
);

CustomWithClone.Clone = (data) => {
  const { props, message, child } = data;
  return React.cloneElement(child, {
    ...props,
    message,
    label: "Mon message :",
  });
};

const defaultProps = {
  message: "message erreur",
  messageType: "error",
  forceDisplayMessage: false,
};

const renderFieldForm = ({
  children = "",
  props = defaultProps,
  customProps = {},
}) =>
  render(
    <FieldForm {...props} {...customProps}>
      {children}
      <FieldError message={props.message} messageType={props.messageType} />
    </FieldForm>,
  );

const expectBySelector = (container, selector, not = false) => {
  const element = container.querySelector(selector);
  not
    ? expect(element).not.toBeInTheDocument()
    : expect(element).toBeInTheDocument();
  return element;
};

const expectText = (getByText, text) =>
  expect(getByText(RegExp(text))).toBeInTheDocument();

describe("Render <FieldForm/>", () => {
  it("Should render empty children", () => {
    const { asFragment, container } = renderFieldForm({});
    expectBySelector(container, ".md-10");
    expect(asFragment()).toMatchSnapshot();
  });

  it("Should render no JSX children", () => {
    const { asFragment, container, getByText } = renderFieldForm({
      children: "testtest",
    });
    expectBySelector(container, ".md-10");
    expectText(getByText, "testtest");
    expect(asFragment()).toMatchSnapshot();
  });

  it("Should render simple JSX children without type.Clone", () => {
    const Custom = () => <p>Simple JSX</p>;
    const { asFragment, container, getByText } = renderFieldForm({
      children: <Custom />,
    });
    expectBySelector(container, ".md-10");
    expectText(getByText, "Simple JSX");
    expect(asFragment()).toMatchSnapshot();
  });

  it("Should render custom JSX children with type.Clone and forceDisplayMessage is false", () => {
    const { asFragment, container, getByText } = renderFieldForm({
      children: <CustomWithClone />,
    });
    expectBySelector(container, ".md-10");
    expectText(getByText, "Mon message :");
    expect(asFragment()).toMatchSnapshot();
  });

  it("Should render custom JSX children and display error message with type.Clone and forceDisplayMessage is true", () => {
    const { asFragment, container, getByText } = renderFieldForm({
      customProps: { forceDisplayMessage: true },
      children: <CustomWithClone />,
    });
    expectBySelector(container, ".md-10");
    expectText(getByText, "Mon message : message erreur");
    expect(asFragment()).toMatchSnapshot();
  });
});

describe("Render <FieldForm/> with <CustomWithClone />", () => {
  it("Should render with message error displayed when forceDisplayMessage true", () => {
    const { asFragment, container, getByText } = renderFieldForm({
      customProps: { forceDisplayMessage: true },
      children: <CustomWithClone />,
    });
    expectBySelector(container, ".af-form__message--error");
    expectText(getByText, "Mon message : message erreur");
    expect(asFragment()).toMatchSnapshot();
  });
});

describe("Render <FieldForm/> with <FieldInput />", () => {
  it("Should render with message displayed when forceDisplayMessage true", () => {
    const { asFragment, container, getByText } = renderFieldForm({
      customProps: {
        forceDisplayMessage: true,
        messageType: "success",
        message: "bravo",
      },
      children: (
        <FieldInput className="sam">
          <CustomWithClone />
        </FieldInput>
      ),
    });
    expectBySelector(container, ".af-form__message--error");
    expectBySelector(container, ".sam.sam--success");
    const fieldErrorMessage = expectBySelector(
      container,
      ".af-form__error-text",
    );
    expectText(within(fieldErrorMessage).getByText, "bravo");
    expectText(getByText, "Mon message : bravo");
    expect(asFragment()).toMatchSnapshot();
  });

  it("Should render with display error message when forceDisplayMessage true", () => {
    const { asFragment, container, getByText } = renderFieldForm({
      customProps: { forceDisplayMessage: true },
      children: (
        <FieldInput className="sam">
          <CustomWithClone />
        </FieldInput>
      ),
    });
    expectBySelector(container, ".af-form__message--error");
    expectBySelector(container, ".sam.sam--error");
    const fieldErrorMessage = expectBySelector(
      container,
      ".af-form__error-text",
    );
    expectText(within(fieldErrorMessage).getByText, "message erreur");
    expectText(getByText, "Mon message : message erreur");
    expect(asFragment()).toMatchSnapshot();
  });
});

describe("Render <FieldForm/> with <MyField />", () => {
  const MyField = (props) => (
    <input
      type="text"
      id="name"
      name="name"
      aria-label="input-name"
      {...props}
    />
  );

  const renderCompleteFieldForm = (customProps) =>
    renderFieldForm({
      customProps,
      children: (
        <Fragment>
          <FieldInput className="sam">
            <MyField />
          </FieldInput>
          <HelpMessage message="un peu d'aide ?" />
        </Fragment>
      ),
    });

  it("Should render with FieldInput and FieldError displayed and HelpMessage hidden when forceDisplayMessage true ", () => {
    const { asFragment, container, getByLabelText } = renderCompleteFieldForm({
      forceDisplayMessage: true,
    });

    expect(getByLabelText("input-name")).toBeInTheDocument();
    expectBySelector(container, ".af-form__message--error");
    expectBySelector(container, ".sam.sam--error");
    const fieldErrorMessage = expectBySelector(
      container,
      ".af-form__error-text",
    );
    expectText(within(fieldErrorMessage).getByText, "message erreur");
    expect(asFragment()).toMatchSnapshot();
  });

  it("Should render with FieldInput and HelpMessage displayed and FieldError hidden when message is null", () => {
    const { asFragment, getByText, container, getByLabelText } =
      renderCompleteFieldForm({ message: null });
    expectBySelector(container, ".af-form__message--error", true);
    expect(getByLabelText("input-name")).toBeInTheDocument();
    expectText(getByText, "un peu d'aide ?");
    expect(asFragment()).toMatchSnapshot();
  });
});

const defaultParams = {
  children: undefined,
  wrapper: null,
  message: null,
  messageType: "",
};

describe("renderedChildren", () => {
  it("Should return null when children undefined", () => {
    const result = renderedChildren(defaultParams);
    expect(result).toBeNull();
  });

  it("Should return empty array when children []", () => {
    const result = renderedChildren({ ...defaultParams, children: [] });
    expect(result).toEqual([]);
  });

  it("Should return the same child when child has no props", () => {
    const Child = () => <p>child</p>;
    const result = renderedChildren({
      ...defaultParams,
      children: <Child />,
    });
    expect(result[0].props).toEqual({});
  });

  it("Should return the only childs with type when child has props", () => {
    const ChildWithProps = ({ className, id }) => (
      <p className={className} id={id}>
        child
      </p>
    );
    const result = renderedChildren({
      ...defaultParams,
      children: [
        null,
        undefined,
        <p>test</p>,
        <ChildWithProps className="hello" id="world" />,
        "str",
        true,
        false,
        <hr />,
        <p>{undefined}</p>,
        <p>{null}</p>,
      ],
    });

    expect(result.length).toEqual(6);
    expect(result[0].type).toEqual("p");
    expect(result[0].props).toEqual({ children: ["test"] });
    expect(result[1].props).toEqual({ className: "hello", id: "world" });
    expect(result[2]).toEqual("str");
    expect(result[3].type).toEqual("hr");
  });
});

describe("getInfo", () => {
  const memoryMessage = "memory message";
  const memoryMessageType = "memory messageType";
  const memory = {
    message: memoryMessage,
    messageType: memoryMessageType,
  };

  const message = "message erreur";
  const messageType = "error";
  it.each`
    forceDisplayMessage | hasLostFocusOnce | hasFocus | hasChange | expectedMessage  | expectedMessageType
    ${false}            | ${false}         | ${false} | ${false}  | ${""}            | ${""}
    ${false}            | ${false}         | ${false} | ${true}   | ${message}       | ${messageType}
    ${false}            | ${false}         | ${true}  | ${false}  | ${""}            | ${""}
    ${false}            | ${false}         | ${true}  | ${true}   | ${""}            | ${""}
    ${false}            | ${true}          | ${false} | ${false}  | ${message}       | ${messageType}
    ${false}            | ${true}          | ${false} | ${true}   | ${message}       | ${messageType}
    ${false}            | ${true}          | ${true}  | ${false}  | ${memoryMessage} | ${memoryMessageType}
    ${false}            | ${true}          | ${true}  | ${true}   | ${memoryMessage} | ${memoryMessageType}
    ${true}             | ${false}         | ${false} | ${false}  | ${message}       | ${messageType}
    ${true}             | ${false}         | ${false} | ${true}   | ${message}       | ${messageType}
    ${true}             | ${false}         | ${true}  | ${false}  | ${memoryMessage} | ${memoryMessageType}
    ${true}             | ${false}         | ${true}  | ${true}   | ${memoryMessage} | ${memoryMessageType}
    ${true}             | ${true}          | ${false} | ${false}  | ${message}       | ${messageType}
    ${true}             | ${true}          | ${false} | ${true}   | ${message}       | ${messageType}
    ${true}             | ${true}          | ${true}  | ${false}  | ${memoryMessage} | ${memoryMessageType}
    ${true}             | ${true}          | ${true}  | ${true}   | ${memoryMessage} | ${memoryMessageType}
  `(
    "Should return expectedMessage : '$expectedMessage' and expectedMessageType : '$expectedMessageType' when forceDisplayMessage: $forceDisplayMessage, hasLostFocusOnce: $hasLostFocusOnce, hasFocus: $hasFocus , hasChange: $hasChange",
    ({
      forceDisplayMessage,
      hasLostFocusOnce,
      hasFocus,
      hasChange,
      expectedMessage,
      expectedMessageType,
    }) => {
      expect(
        getInfo({
          forceDisplayMessage,
          hasLostFocusOnce,
          hasFocus,
          hasChange,
          memory,
          message,
          messageType,
        }),
      ).toEqual({ message: expectedMessage, messageType: expectedMessageType });
    },
  );
});
