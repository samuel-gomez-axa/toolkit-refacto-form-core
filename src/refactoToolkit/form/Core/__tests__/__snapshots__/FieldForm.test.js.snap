// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Render <FieldForm/> Should render custom JSX children and display error message with type.Clone and forceDisplayMessage is true 1`] = `
<DocumentFragment>
  <div
    class="md-10"
  >
    <p>
      Mon message : message erreur
    </p>
    <small
      class="af-form__message af-form__message--error"
    >
      <span
        class="glyphicon glyphicon-exclamation-sign"
      />
      <span
        class="af-form__error-text"
      >
        message erreur
      </span>
    </small>
  </div>
</DocumentFragment>
`;

exports[`Render <FieldForm/> Should render custom JSX children with type.Clone and forceDisplayMessage is false 1`] = `
<DocumentFragment>
  <div
    class="md-10"
  >
    <p>
      Mon message : 
    </p>
  </div>
</DocumentFragment>
`;

exports[`Render <FieldForm/> Should render empty children 1`] = `
<DocumentFragment>
  <div
    class="md-10"
  />
</DocumentFragment>
`;

exports[`Render <FieldForm/> Should render no JSX children 1`] = `
<DocumentFragment>
  <div
    class="md-10"
  >
    testtest
  </div>
</DocumentFragment>
`;

exports[`Render <FieldForm/> Should render simple JSX children without type.Clone 1`] = `
<DocumentFragment>
  <div
    class="md-10"
  >
    <p>
      Simple JSX
    </p>
  </div>
</DocumentFragment>
`;

exports[`Render <FieldForm/> with <CustomWithClone /> Should render with message error displayed when forceDisplayMessage true 1`] = `
<DocumentFragment>
  <div
    class="md-10"
  >
    <p>
      Mon message : message erreur
    </p>
    <small
      class="af-form__message af-form__message--error"
    >
      <span
        class="glyphicon glyphicon-exclamation-sign"
      />
      <span
        class="af-form__error-text"
      >
        message erreur
      </span>
    </small>
  </div>
</DocumentFragment>
`;

exports[`Render <FieldForm/> with <FieldInput /> Should render with display error message when forceDisplayMessage true 1`] = `
<DocumentFragment>
  <div
    class="md-10"
  >
    <div
      class="sam sam--error"
    >
      <p>
        Mon message : message erreur
      </p>
    </div>
    <small
      class="af-form__message af-form__message--error"
    >
      <span
        class="glyphicon glyphicon-exclamation-sign"
      />
      <span
        class="af-form__error-text"
      >
        message erreur
      </span>
    </small>
  </div>
</DocumentFragment>
`;

exports[`Render <FieldForm/> with <FieldInput /> Should render with message displayed when forceDisplayMessage true 1`] = `
<DocumentFragment>
  <div
    class="md-10"
  >
    <div
      class="sam sam--success"
    >
      <p>
        Mon message : bravo
      </p>
    </div>
    <small
      class="af-form__message af-form__message--error"
    >
      <span
        class="glyphicon glyphicon-exclamation-sign"
      />
      <span
        class="af-form__error-text"
      >
        bravo
      </span>
    </small>
  </div>
</DocumentFragment>
`;

exports[`Render <FieldForm/> with <MyField /> Should render with FieldInput and FieldError displayed and HelpMessage hidden when forceDisplayMessage true  1`] = `
<DocumentFragment>
  <div
    class="md-10"
  >
    <div
      class="sam sam--error"
    >
      <input
        aria-label="input-name"
        id="name"
        name="name"
        type="text"
      />
    </div>
    <small
      class="af-form__message af-form__message--error"
    >
      <span
        class="glyphicon glyphicon-exclamation-sign"
      />
      <span
        class="af-form__error-text"
      >
        message erreur
      </span>
    </small>
  </div>
</DocumentFragment>
`;

exports[`Render <FieldForm/> with <MyField /> Should render with FieldInput and HelpMessage displayed and FieldError hidden when message is null 1`] = `
<DocumentFragment>
  <div
    class="md-10"
  >
    <div
      class="sam"
    >
      <input
        aria-label="input-name"
        id="name"
        name="name"
        type="text"
      />
    </div>
    <small
      class="af-form__help"
    >
      un peu d'aide ?
    </small>
  </div>
</DocumentFragment>
`;
